addons/libs/dialog.lua

```diff
@@ -0,0 +1,158 @@
+-- This library was written to help find the ID of a known
+-- action message corresponding to an entry in the dialog tables.
+-- While the IDs can be collected in-game, they occasionally
+-- change and would otherwise need to be manually updated.
+-- It can also be used to find and decode an entry given the ID.
+
+-- Common parameters:
+-- dat: The entire content of the zone dialog DAT file
+-- i.e. local dat = io.open('path/to/dialog/DAT', 'rb'):read('*a')
+-- entry: The string you are looking for. If you do not know the
+-- entire string, use dev_find_substring.
+
+local xor = require('bit').bxor
+require('pack')
+local string = require('string')
+local unpack = string.unpack
+local pack = string.pack
+local find = string.find
+
+local dialog = {}
+
+local function decode(int)
+    return xor(int, 0x80808080)
+end
+local encode = decode
+
+local floor = require('math').floor
+local function binary_search(pos, dat, n)
+    local l, r, m = 1, n
+    while l < r do
+        m = floor((l+r)/2)
+        if decode(unpack('<I', dat, 1+4*m)) < pos then
+        -- offset given by mth ID < offset to string
+            l = m + 1
+        else
+            r = m
+        end
+    end
+    return l-2 -- we want the index to the left of where "pos" would be placed
+end
+
+local function plain_text_gmatch(text, substring, n)
+    n = n or 1
+    return function()
+        local pos = find(text, substring, n, true)
+        if pos then n = pos + 1 end
+        return pos
+    end
+end
+
+function dialog.dev_get_offset(dat, id) -- sanity check function
```

Comment:
I think it would be nicer to have all the development functions in a `dialog.dev.*` subtable. That said, isn't the entire library meant as a dev tool? Or is it meant to be used productively?

Reply:
That was originally the plan, but there ended up only being two `dev_*` functions, and the table seemed like a waste. I marked those functions to emphasize that while the others might somehow be useful in an addon, those definitely shouldn't appear: `dev_get_offset` is meant to help alongside a hex editor, and `dev_find_entry` is slow and ambiguous.
There is one case that I can think of where an addon could use the functions outside of development: the messages pointwatch and boxdestroyer use are unique, so the result from calling `get_ids_matching_entry` when you enter one of those zones could be used in place of manually updating the message IDs.

Reply:
I still think a sub table would make sense, even for just those two. May also add more in the future, I could see a library like this being helpful for a bunch of things.